src/main/kotlin/org/rust/toml/CargoTomlKeyReferenceContributor.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml
+
+import com.intellij.openapi.util.TextRange
+import com.intellij.patterns.PsiElementPattern
+import com.intellij.psi.*
+import com.intellij.util.ProcessingContext
+import org.rust.cargo.project.model.cargoProjects
+import org.rust.lang.core.or
+import org.rust.lang.core.psi.RsFile
+import org.rust.lang.core.psiElement
+import org.rust.lang.core.with
+import org.toml.lang.psi.TomlKey
+import org.toml.lang.psi.TomlTable
+import org.toml.lang.psi.TomlTableHeader
+
+class CargoTomlKeyReferenceContributor : PsiReferenceContributor() {
+
+    override fun registerReferenceProviders(registrar: PsiReferenceRegistrar) {
+        registrar.registerReferenceProvider(dependencyKeyPattern or specificDependencyKeyPattern,
+            CargoDependencyReferenceProvider())
+    }
+
+    companion object {
+
+        private const val TOML_KEY_CONTEXT_NAME = "key"
+
+        private val dependencyKeyPattern: PsiElementPattern.Capture<TomlKey> = psiElement<TomlKey>()
+            .withSuperParent(2, psiElement<TomlTable>()
+                .withChild(psiElement<TomlTableHeader>()
+                    .with("dependenciesCondition") { e ->
+                        if (e !is TomlTableHeader) return@with false
+                        e.names.lastOrNull()?.isDependencyKey == true
+                    }
+                )
+            )
+
+        private val specificDependencyKeyPattern: PsiElementPattern.Capture<TomlKey> = psiElement<TomlKey>(TOML_KEY_CONTEXT_NAME)
+            .withParent(psiElement<TomlTableHeader>()
+                .with("specificDependencyCondition") { e, context ->
+                    val key = context?.get(TOML_KEY_CONTEXT_NAME) ?: return@with false
+                    if (e !is TomlTableHeader) return@with false
```

Comment:
This check better be replaced with `if (e == null) return@with false`
And maybe the whole lambda body should be moved to the separate function to avoid so many local returns?

Reply:
Hm, I was sure that I need a cast here but I don't know why.
Really, we don't such check at all because `e` is always not null